### PR TITLE
fix: correct suggestion arg for impl trait

### LINKED ISSUE #119775: Diagnostic for `CandidateSource::Trait` take receiver as first arg mistakely
### Code

```Rust
struct A {}

trait M {
    fn new(a: Self);
}
impl M for A {
    fn new(a: Self) {
        todo!()
    }
}

fn main() {
    let a = A {};
    a.new();
}
```

### Current output

```Shell
error[E0599]: no method named `new` found for struct `A` in the current scope
  --> src/main.rs:28:7
   |
15 | struct A {}
   | -------- method `new` not found for this struct
...
28 |     a.new();
   |     --^^^--
   |     | |
   |     | this is an associated function, not a method
   |     help: use associated function syntax instead: `A::new()`
   |
   = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: the candidate is defined in the trait `M`
  --> src/main.rs:18:5
   |
18 |     fn new(a: Self);
   |     ^^^^^^^^^^^^^^^^
   = help: items from traits can only be used if the trait is implemented and in scope
```

### Desired output

```Shell
A::new(a)
```

### Rationale and extra context

We had solved `CandidateSource::Impl` case in https://github.com/rust-lang/rust/pull/118502, could do the same for `CandidateSource::Trait`

https://github.com/rust-lang/rust/blob/be00c5a9b89161b7f45ba80340f709e8e41122f9/compiler/rustc_hir_typeck/src/method/suggest.rs#L1590-L1606

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.76.0-nightly (ba7c7a301 2023-11-13)
binary: rustc
commit-hash: ba7c7a301984967c8c13adb580ef9b86ba706a83
commit-date: 2023-11-13
host: x86_64-unknown-linux-gnu
release: 1.76.0-nightly
LLVM version: 17.0.4
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Young-Flash**
follow up https://github.com/rust-lang/rust/pull/118502, close https://github.com/rust-lang/rust/issues/119775

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 1]
```diff
@@ -1587,23 +1587,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
         };
         if let SelfSource::MethodCall(_) = source {
-            let first_arg = if let Some(CandidateSource::Impl(impl_did)) = static_candidates.get(0)
-                && let Some(assoc) = self.associated_value(*impl_did, item_name)
-                && assoc.kind == ty::AssocKind::Fn
-            {
+            let first_arg = static_candidates.get(0).and_then(|candidate_source| {
+                let (assoc_did, impl_ty) = match candidate_source {
+                    CandidateSource::Impl(impl_did) => {
+                        (*impl_did, self.tcx.type_of(*impl_did).instantiate_identity())
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1593]: Why don't we use the `rcvr_ty` for `CandidateSource::Impl`, too?

Comment (Young-Flash) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:1593]: given the following:
```rust
struct B<T> {
    _b: T
}
impl<T> B<T> {
    fn hello(_a: i32) {}
    fn test(_a: Self, _b: i32) {}
}

fn main() {
    let _b = B {_b: ""};
    _b.test(1);
}
```
`rcvr_ty`: `B<&str>`
`first_ty`: `B<T>`
`self.tcx.type_of(*impl_did).instantiate_identity()`: `B<T>`

Comment (Young-Flash) [reply compiler/rustc_hir_typeck/src/method/suggest.rs:1593]: anything else here can improve?

**Comment by fmease** [comment]
Sorry for the delay, I will review this tomorrow :)

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 2]
```diff
@@ -1587,23 +1587,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             self.ty_to_value_string(rcvr_ty.peel_refs())
         };
         if let SelfSource::MethodCall(_) = source {
-            let first_arg = if let Some(CandidateSource::Impl(impl_did)) = static_candidates.get(0)
-                && let Some(assoc) = self.associated_value(*impl_did, item_name)
-                && assoc.kind == ty::AssocKind::Fn
-            {
+            let first_arg = static_candidates.get(0).and_then(|candidate_source| {
+                let (assoc_did, impl_ty) = match candidate_source {
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1591]:
```suggestion
                let (assoc_did, self_ty) = match candidate_source {
```

“`impl_ty`” doesn't feel right for `CandidateSource::Trait`.

File: compiler/rustc_hir_typeck/src/method/suggest.rs [hunk 3]
```diff
@@ -1587,23 +1587,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+
+                // for CandidateSource::Impl, `Self` will be instantiated to a concrete type
+                // but for CandidateSource::Trait, `Self` is still `Self`
                 let sig = self.tcx.fn_sig(assoc.def_id).instantiate_identity();
                 sig.inputs().skip_binder().get(0).and_then(|first| {
-                    let impl_ty = self.tcx.type_of(*impl_did).instantiate_identity();
                     // if the type of first arg is the same as the current impl type, we should take the first arg into assoc function
-                    if first.peel_refs() == impl_ty {
+                    let first_ty = first.peel_refs();
+                    if first_ty == impl_ty || first_ty == self.tcx.types.self_param {
```
Comment (fmease) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1609]:
```suggestion
                    if first_ty == self_ty || first_ty == self.tcx.types.self_param {
```

**Review by fmease — APPROVED** [review]
Thanks! Minor suggestion, then we're good to go!

**Pull request merged**